src/collections.rs

```diff
@@ -0,0 +1,236 @@
+use async_std::channel::{Receiver, Sender};
+use async_std::sync::Mutex;
+use std::collections::VecDeque;
+use std::ops::Deref;
+use tracing::error;
+
+/// Blocking async double-ended queue with interior mutability.
+pub struct BlockingDeque<T> {
+    queue: Mutex<VecDeque<T>>,
+    control: (Sender<()>, Receiver<()>),
+}
+
+impl<T> BlockingDeque<T> {
+    /// Creates a new queue.
+    ///
+    /// The created queue has space to hold at most `cap` records at a time.
+    ///
+    /// # Panics
+    ///
+    /// Capacity must be a positive number. If `cap` is zero, this function will panic.
+    pub fn new(cap: usize) -> Self {
+        let queue = Mutex::new(VecDeque::new());
+        let control = async_std::channel::bounded(cap);
+        Self { queue, control }
+    }
+
+    /// Appends an element to the back of the deque.
+    ///
+    /// If the queue is full, this method waits until there is space for a value.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use fluvio_future::collections::BlockingDeque;
+    /// fluvio_future::task::run(async {
+    ///     let queue = BlockingDeque::new(2);
+    ///     assert!(queue.is_empty());
+    ///     queue.push_back(1).await;
+    ///     queue.push_back(1).await;
+    ///     assert_eq!(queue.len(), 2);
+    /// });
+    /// ```
+    pub async fn push_back(&self, value: T) {
+        let _ = self.control.0.send(()).await; // drop the error because we own both sides
+        self.queue.lock().await.push_back(value);
```

Comment:
Instead of trying to lock, which could cause deadlock, it would be better to use `try_lock.`  if the lock fails, then. perform a wait loop with a timeout.  I think that would avoid using channel